core/src/block.rs

```diff
@@ -95,74 +101,74 @@ impl Chain {
+            return None;
         }
-    }
-
-    const fn is_exhausted(&self) -> bool {
-        self.pos_front > self.pos_back
-    }
-}
 
-impl<'itm> Iterator for ChainIterator<'itm> {
-    type Item = MapRef<'itm, u64, VersionedCommittedBlock>;
-    fn next(&mut self) -> Option<Self::Item> {
-        if !self.is_exhausted() {
-            let val = self.chain.blocks.get(&self.pos_front);
-            self.pos_front += 1;
-            return val;
+        let mut index: isize = (self.chain_mutex_guard.blocks.len() - 1)
+            .try_into()
+            .expect("we don't have more than 2^63 blocks");
+        while index >= 0 {
+            if self.chain_mutex_guard.blocks[index as usize]
+                .header()
+                .previous_block_hash
+                == hash
+            {
+                return Some(&self.chain_mutex_guard.blocks[index as usize..]);
+            }
+            index -= 1;
         }
         None
     }
-
-    fn nth(&mut self, n: usize) -> Option<Self::Item> {
-        self.pos_front += n as u64;
-        self.next()
-    }
-
-    fn last(mut self) -> Option<Self::Item> {
-        self.pos_front = self.chain.len() as u64;
-        self.chain.blocks.get(&self.pos_front)
-    }
-
-    fn count(self) -> usize {
-        #[allow(clippy::cast_possible_truncation)]
-        let count = (self.chain.len() as u64 - (self.pos_front - 1)) as usize;
-        count
-    }
-
-    fn size_hint(&self) -> (usize, Option<usize>) {
-        #[allow(clippy::cast_possible_truncation)]
-        let height = (self.chain.len() as u64 - (self.pos_front - 1)) as usize;
-        (height, Some(height))
-    }
-}
-
-impl<'itm> DoubleEndedIterator for ChainIterator<'itm> {
-    fn next_back(&mut self) -> Option<Self::Item> {
-        if !self.is_exhausted() {
-            let val = self.chain.blocks.get(&self.pos_back);
-            self.pos_back -= 1;
-            return val;
+    /// Return a slice of all blocks in the `Chain` after and including
+    /// the block at the provided height.
+    #[allow(clippy::expect_used, clippy::unwrap_in_result)]
+    pub fn blocks_from_height(&self, height: u64) -> Option<&[VersionedCommittedBlock]> {
+        if height
+            >= self.chain_mutex_guard.base_block_height + self.chain_mutex_guard.blocks.len() as u64
+        {
+            return None;
         }
-        None
+        Some(
+            &self.chain_mutex_guard.blocks
+                [height.saturating_sub(1).try_into().expect("fits in usize")..],
+        )
     }
-
-    fn nth_back(&mut self, n: usize) -> Option<Self::Item> {
-        self.pos_back -= n as u64;
-        self.next_back()
+    /// Return a slice of all blocks in the `Chain`.
+    pub fn all_blocks(&self) -> &[VersionedCommittedBlock] {
+        self.chain_mutex_guard.blocks.as_slice()
+    }
+    /// Return the last block in the `Chain`.
+    pub fn last_block_iter(&self) -> Option<&VersionedCommittedBlock> {
```

Comment:
I think `_iter` suffix is confusing, because this function returns `Option`, not an iterator

Reply:
yes